File: ios/Superwall.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(Superwall, RCTEventEmitter)

RCT_EXTERN_METHOD(initPaywall:(NSString)superwallApiKey revenueCatApiKey:(NSString)revenueCatApiKey appUserID:(NSString)appUserID)

RCT_EXTERN_METHOD(trigger:(NSString)campaignName resolver: (RCTPromiseResolveBlock)resolve rejecter: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(reset)

RCT_EXTERN_METHOD(identify:(NSString)appUserID)

RCT_EXTERN_METHOD(setUserAttributes:(NSDictionary*)userAttributes)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: ios/Superwall.swift
import Foundation
import Paywall

@objc(Superwall)
class Superwall: RCTEventEmitter {

    public static var emitter: RCTEventEmitter!

      override init() {
        super.init()
        Superwall.emitter = self
      }

      open override func supportedEvents() -> [String] {
        ["superwallAnalyticsEvent"]
      }


    @objc(initPaywall:revenueCatApiKey:appUserID:)
    func initPaywall(superwallApiKey:String, revenueCatApiKey:String, appUserID:String) -> Void {
        PaywallService.initPaywall(superwallApiKey: superwallApiKey, revenueCatApiKey: revenueCatApiKey, appUserID: appUserID)
    }


    @objc(trigger:resolver:rejecter:)
    func trigger(campaignName:String, resolver: @escaping RCTPromiseResolveBlock,  rejecter: @escaping RCTPromiseRejectBlock) -> Void {
        Paywall.trigger(
          event: campaignName,
          params: nil,
          onSkip: { error in
            rejecter("error", error?.description, nil)
          },
          onPresent: nil,
          onDismiss: { didPurchase, productId, paywallInfo in
              if didPurchase {
                  resolver(["didPurchase": true, "productId": productId])
              } else {
                resolver(["didPurchase": false, "productId": nil])
              }
            }
        )
    }

    @objc
    func reset() -> Void {
        PaywallService.reset();
    }

    @objc(identify:)
    func identify(appUserID:String) -> Void {
        PaywallService.identify(appUserID: appUserID);
    }

}
